File: Workers/aidp/retrieval.py
"""Hybrid retrieval, for the analyse stage.

Mirrors src/lib/ingest/retrieval.ts. The two have to agree, because the app
embeds questions and this embeds reference clauses against the same vectors —
divergence would show up as quietly worse results rather than as an error.

Dense vectors plus Postgres full-text, fused with Reciprocal Rank Fusion. Both
halves matter here more than anywhere else in the system: the analyse stage has
to tell "this requirement is genuinely unaddressed" from "retrieval missed it",
and that rests entirely on recall. Dense search is soft on exactly the tokens
these documents turn on — "TLS 1.2", "RPO", "snake_case", "MFA".
"""

from __future__ import annotations

from dataclasses import dataclass

import psycopg

from . import db
from .ai import embeddings
from .config import get_config

# Reciprocal Rank Fusion constant, from the original paper.
RRF_K = 60


@dataclass
class Candidate:
    chunk_id: str
    heading_path: str
    text: str
    page_start: int | None
    score: float
    vector_rank: int | None
    lexical_rank: int | None
    # Where the passage came from. "figure" means it is a *model's reading of a
    # diagram*, not text lifted off the page — which the judge is told, and
    # which the guards in analyse.py refuse to let decide a verdict alone.
    source_kind: str = "clause"
    # The row this chunk was derived from. For a figure that is the figure id,
    # which is what lets a finding render the diagram beside the model's reading
    # of it — verification at the point of use rather than at ingest.
    source_id: str | None = None

    @property
    def is_generated(self) -> bool:
        return self.source_kind == "figure"

    @property
    def excerpt(self) -> str:
        """The passage as a reviewer will read it.

        Prose can have its line breaks collapsed — a PDF wraps mid-sentence and
        those breaks mean nothing. A table's rows are the opposite: which value
        sits in which column *is* the content. Collapsing them fuses the last
        cell of one row onto the first of the next ("Example Table",
        "CustomerOrders Column"), and the result cannot be untangled afterwards
        because the join is indistinguishable from a two-word cell.
        """
        if self.source_kind in ("table", "table_row"):
            body = "\n".join(
                " ".join(line.split()) for line in self.text.splitlines() if line.strip()
            )
        else:
            body = " ".join(self.text.split())
        return body if len(body) <= 700 else body[:700] + "…"


_SEARCH = """
WITH dense AS (
    SELECT c."id",
           ROW_NUMBER() OVER (ORDER BY e."vector" <=> %(vector)s::vector) AS rank
      FROM "chunk" c
      JOIN "embedding" e ON e."chunkId" = c."id" AND e."model" = %(model)s
     WHERE c."organisationId" = %(org)s
       AND c."documentId" = %(doc)s
     ORDER BY e."vector" <=> %(vector)s::vector
     LIMIT %(pool)s
),
lexical AS (
    SELECT c."id",
           ROW_NUMBER() OVER (
             ORDER BY ts_rank_cd(
               aidp_chunk_vector(c."headingPath", c."text"), q.query
             ) DESC
           ) AS rank
      FROM "chunk" c
     -- Both functions are defined in the migration, not here, because
     -- src/lib/ingest/retrieval.ts runs the same search over the same index and
     -- the two had already drifted once. `aidp_search_query` ORs the lexemes
     -- (a clause AND-ed matches nothing) and drops the corpus-wide filler that
     -- made the query match most of a document; `aidp_chunk_vector` weights the
     -- heading path above the body and is what the GIN index is built on.
     --
     -- Never write a percent sign anywhere in this string, comments included.
     -- psycopg scans the whole statement for placeholders, so a stray one
     -- either fails as a truncated placeholder or, worse, parses as a real
     -- named one with no matching parameter. Both fail at execute time rather
     -- than at import, so a comment can break retrieval in production. Spell
     -- the word out instead.
     CROSS JOIN (SELECT aidp_search_query(%(query)s) AS query) q
     WHERE c."organisationId" = %(org)s
       AND c."documentId" = %(doc)s
       AND aidp_chunk_vector(c."headingPath", c."text") @@ q.query
     LIMIT %(pool)s
),
fused AS (
    SELECT COALESCE(d."id", l."id") AS id,
           COALESCE(1.0 / (%(k)s + d.rank), 0)
         + COALESCE(1.0 / (%(k)s + l.rank), 0) AS score,
           d.rank AS vector_rank,
           l.rank AS lexical_rank
      FROM dense d
      FULL OUTER JOIN lexical l ON l."id" = d."id"
)
SELECT c."id", c."headingPath", c."text", c."pageStart", c."sourceKind", c."sourceId",
       f.score::float8 AS score, f.vector_rank, f.lexical_rank
  FROM fused f
  JOIN "chunk" c ON c."id" = f.id
 ORDER BY f.score DESC
 LIMIT %(limit)s
"""


def embed_query(query: str) -> str:
    """A query as a pgvector literal.

    Exposed so one clause can be embedded once and the vector shared between
    passage retrieval and the decision lookup, rather than paying the provider
    twice for the same string on every clause of every run.
    """
    return embeddings.to_pgvector(embeddings.embed_all([query], "query")[0])


def search_document(
    conn: psycopg.Connection,
    *,
    organisation_id: str,
    document_id: str,
    query: str,
    limit: int = 8,
    vector: str | None = None,
) -> list[Candidate]:
    """Best passages in one document for one query.

    `organisation_id` is required and lands in the predicate even though
    `document_id` alone would be selective enough. Belt and braces: a document
    id arriving from a job payload is not a capability check, and the tenant
    filter should be present on every path that reads chunks.
    """
    if not query.strip():
        return []

    cfg = get_config()
    if vector is None:
        vector = embed_query(query)

    # HNSW discards non-matching rows *after* walking the graph, so a filter as
    # tight as a single document can starve the result set. pgvector 0.8 keeps
    # scanning until k survivors are found.
    with conn.cursor() as cur:
        cur.execute("SET LOCAL hnsw.iterative_scan = 'relaxed_order'")

    rows = db.query(
        conn,
        _SEARCH,
        {
            "vector": vector,
            "model": cfg.embedding_model,
            "org": organisation_id,
            "doc": document_id,
            "query": query,
            # Fuse from a wider pool than we return, or the two rankings barely
            # overlap and fusion has nothing to work with.
            "pool": max(limit * 4, 32),
            "limit": limit,
            "k": RRF_K,
        },
    )

    return [
        Candidate(
            chunk_id=r["id"],
            heading_path=r["headingPath"],
            text=r["text"],
            page_start=r["pageStart"],
            score=float(r["score"] or 0),
            vector_rank=r["vector_rank"],
            lexical_rank=r["lexical_rank"],
            source_kind=r["sourceKind"],
            source_id=r["sourceId"],
        )
        for r in rows
    ]


def clause_query(statement: str, requirements: list[str], title: str | None) -> str:
    """Turn a reference clause into a search query.

    The requirements bullets carry most of the useful vocabulary — they are
    where the concrete nouns live ("dead-letter queue", "TLS 1.2", "recovery
    point objective"), and those are what the other document will actually say.
    The statement alone is usually too abstract to match anything.
    """
    parts = [p for p in [title, statement, *requirements] if p]
    query = " ".join(parts)
    # plainto_tsquery copes with long input, but the embedding call does not
    # need the whole thing and the tail adds nothing.
    return query[:2000]

File: AIDP/src/lib/ingest/retrieval.ts
import { Prisma } from "../../../generated/prisma/client";
import { prisma } from "@/lib/prisma";
import { requireMembership } from "./org";

/**
 * The only way to search the corpus.
 *
 * Two things are enforced here rather than left to callers:
 *
 *  1. **The tenant filter.** `organisationId` is a required argument and lands
 *     in the SQL predicate, not in a prompt. Membership is checked before the
 *     query runs. There is deliberately no variant of this function without it.
 *
 *  2. **Hybrid retrieval.** Dense vectors plus Postgres full-text, fused. The
 *     analyse stage has to tell "this requirement is genuinely unaddressed"
 *     from "retrieval missed it", and that rests on recall. Dense search alone
 *     is soft on exactly the tokens that matter most in these documents —
 *     "TLS 1.2", "RPO", "snake_case", "MFA".
 */

export type Hit = {
  chunkId: string;
  documentId: string;
  documentTitle: string;
  headingPath: string;
  sourceKind: string;
  text: string;
  pageStart: number | null;
  score: number;
  vectorRank: number | null;
  lexicalRank: number | null;
};

export type SearchOptions = {
  userId: string;
  organisationId: string;
  query: string;
  limit?: number;
  /** Narrow to specific documents — how the analyse stage scopes to one submission. */
  documentIds?: string[];
  /**
   * Classifications the caller may see. Data Standards §8.1 defines the tiers;
   * omitting this returns everything the organisation holds, which is only
   * right for an internal consultant view.
   */
  sensitivities?: string[];
};

/** Reciprocal Rank Fusion constant. 60 is the value from the original paper. */
const RRF_K = 60;

/**
 * Embed text.
 *
 * Asymmetric on providers that support it: a question and a passage are
 * embedded differently, and getting `input_type` backwards degrades retrieval
 * quietly rather than loudly. Kept in step with Workers/aidp/ai/embeddings.py.
 *
 * `kind` is which side of that asymmetry the caller is on. Search queries are
 * "query"; anything stored to be *found* by a query — a passage, or a standing
 * decision in the register — is "passage".
 */
export async function embedText(
  text: string,
  kind: "query" | "passage" = "query",
): Promise<number[]> {
  const provider = (process.env.EMBEDDING_PROVIDER ?? "gemini").toLowerCase();
  const model = process.env.EMBEDDING_MODEL ?? "gemini-embedding-001";
  const dims = Number(process.env.EMBEDDING_DIMS ?? 1024);

  if (provider === "gemini") {
    const key = process.env.GEMINI_API_KEY ?? process.env.GOOGLE_API_KEY;
    if (!key) throw new Error("GEMINI_API_KEY is not set");
    const res = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/${model}:embedContent`,
      {
        method: "POST",
        headers: { "x-goog-api-key": key, "content-type": "application/json" },
        body: JSON.stringify({
          model: `models/${model}`,
          content: { parts: [{ text }] },
          taskType: kind === "query" ? "RETRIEVAL_QUERY" : "RETRIEVAL_DOCUMENT",
          outputDimensionality: dims,
        }),
      },
    );
    if (!res.ok) throw new Error(`Gemini embeddings failed: ${res.status}`);
    const json = (await res.json()) as { embedding: { values: number[] } };
    return json.embedding.values;
  }

  if (provider === "voyage") {
    const key = process.env.VOYAGE_API_KEY;
    if (!key) throw new Error("VOYAGE_API_KEY is not set");
    const res = await fetch("https://api.voyageai.com/v1/embeddings", {
      method: "POST",
      headers: { authorization: `Bearer ${key}`, "content-type": "application/json" },
      body: JSON.stringify({ model, input: [text], input_type: kind === "query" ? "query" : "document" }),
    });
    if (!res.ok) throw new Error(`Voyage embeddings failed: ${res.status}`);
    const json = (await res.json()) as { data: { embedding: number[] }[] };
    return json.data[0]!.embedding;
  }

  if (provider === "openai") {
    const key = process.env.OPENAI_API_KEY;
    if (!key) throw new Error("OPENAI_API_KEY is not set");
    const res = await fetch("https://api.openai.com/v1/embeddings", {
      method: "POST",
      headers: { authorization: `Bearer ${key}`, "content-type": "application/json" },
      body: JSON.stringify({ model, input: [text], dimensions: dims }),
    });
    if (!res.ok) throw new Error(`OpenAI embeddings failed: ${res.status}`);
    const json = (await res.json()) as { data: { embedding: number[] }[] };
    return json.data[0]!.embedding;
  }

  throw new Error(`Unknown EMBEDDING_PROVIDER: ${provider}`);
}

export function toVectorLiteral(vector: number[]): string {
  return `[${vector.join(",")}]`;
}

export async function search(options: SearchOptions): Promise<Hit[]> {
  const { userId, organisationId, query, limit = 12 } = options;
  await requireMembership(userId, organisationId);
  if (!query.trim()) return [];

  const vector = toVectorLiteral(await embedText(query, "query"));
  const model = process.env.EMBEDDING_MODEL ?? "voyage-3";
  // Fuse from a wider candidate pool than we return, or the two rankings barely
  // overlap and fusion has nothing to work with.
  const pool = Math.max(limit * 4, 40);

  const documentFilter = options.documentIds?.length
    ? Prisma.sql`AND c."documentId" = ANY(${options.documentIds})`
    : Prisma.empty;
  const sensitivityFilter = options.sensitivities?.length
    ? Prisma.sql`AND (c."sensitivity" IS NULL OR c."sensitivity" = ANY(${options.sensitivities}))`
    : Prisma.empty;

  return prisma.$transaction(async (tx) => {
    // HNSW does not pre-filter: it walks the graph and *then* discards rows
    // failing the WHERE clause, so a tight organisation predicate can return
    // fewer than k rows. pgvector 0.8 keeps scanning until k survivors are
    // found. Neon ships 0.8.1, which is what makes this available.
    await tx.$executeRawUnsafe(`SET LOCAL hnsw.iterative_scan = 'relaxed_order'`);

    return tx.$queryRaw<Hit[]>`
      WITH dense AS (
        SELECT c."id",
               ROW_NUMBER() OVER (ORDER BY e."vector" <=> ${vector}::vector) AS rank
          FROM "chunk" c
          JOIN "embedding" e ON e."chunkId" = c."id" AND e."model" = ${model}
         WHERE c."organisationId" = ${organisationId}
           ${documentFilter}
           ${sensitivityFilter}
         ORDER BY e."vector" <=> ${vector}::vector
         LIMIT ${pool}
      ),
      lexical AS (
        SELECT c."id",
               ROW_NUMBER() OVER (
                 ORDER BY ts_rank_cd(
                   aidp_chunk_vector(c."headingPath", c."text"), q.query
                 ) DESC
               ) AS rank
          FROM "chunk" c
         -- Defined in migration 20260807090000_lexical_weighting, and shared
         -- with Workers/aidp/retrieval.py rather than restated here — these two
         -- queries drifted apart once already. aidp_search_query ORs the
         -- lexemes and strips corpus-wide filler ('data', 'must', 'standard');
         -- aidp_chunk_vector weights the heading path above the body and is
         -- the expression chunk_fts_idx is built on, so it must match exactly
         -- or the index goes unused.
         CROSS JOIN (SELECT aidp_search_query(${query}) AS query) q
         WHERE c."organisationId" = ${organisationId}
           AND aidp_chunk_vector(c."headingPath", c."text") @@ q.query
           ${documentFilter}
           ${sensitivityFilter}
         LIMIT ${pool}
      ),
      fused AS (
        SELECT COALESCE(d."id", l."id") AS id,
               COALESCE(1.0 / (${RRF_K} + d.rank), 0)
             + COALESCE(1.0 / (${RRF_K} + l.rank), 0) AS score,
               d.rank AS vector_rank,
               l.rank AS lexical_rank
          FROM dense d
          FULL OUTER JOIN lexical l ON l."id" = d."id"
      )
      SELECT c."id"            AS "chunkId",
             c."documentId"    AS "documentId",
             doc."title"       AS "documentTitle",
             c."headingPath"   AS "headingPath",
             c."sourceKind"    AS "sourceKind",
             c."text"          AS "text",
             c."pageStart"     AS "pageStart",
             f.score::float8   AS "score",
             f.vector_rank::int  AS "vectorRank",
             f.lexical_rank::int AS "lexicalRank"
        FROM fused f
        JOIN "chunk" c ON c."id" = f.id
        JOIN "document" doc ON doc."id" = c."documentId"
       ORDER BY f.score DESC
       LIMIT ${limit}
    `;
  });
}
